bpf/process/bpf_loader.c

```diff
@@ -0,0 +1,140 @@
+// SPDX-License-Identifier: GPL-2.0
+/* Copyright Authors of Cilium */
+
+#include "vmlinux.h"
+#include "api.h"
+#include "bpf_tracing.h"
+#include "bpf_helpers.h"
+#include "lib/common.h"
+#include "lib/hubble_msg.h"
+#include "lib/bpf_events.h"
+
+struct msg_loader {
+	struct msg_common common;
+	struct msg_execve_key current;
+	__u32 pid;
+	__u32 buildid_size;
+	__u32 path_size;
+	char buildid[20];
+	char path[4096];
+	void *pe;
+};
+
+char _license[] __attribute__((section("license"), used)) = "GPL";
+
+struct {
+	__uint(type, BPF_MAP_TYPE_PERCPU_ARRAY);
+	__uint(max_entries, 1);
+	__type(key, __u32);
+	__type(value, struct msg_loader);
+} loader_heap SEC(".maps");
+
+struct {
+	__uint(type, BPF_MAP_TYPE_PERCPU_ARRAY);
+	__uint(max_entries, 1);
+	__type(key, __u32);
+	__type(value, __u64);
+} ids_map SEC(".maps");
+
+struct __perf_event_attr {
+	__u32 type;
+	__u32 size;
+	__u64 config;
+	union {
+		__u64 sample_period;
+		__u64 sample_freq;
+	};
+	__u64 sample_type;
+	__u64 read_format;
+	__u64 bits;
+};
+
+struct {
+	__uint(type, BPF_MAP_TYPE_PERCPU_ARRAY);
+	__uint(max_entries, 1);
+	__type(key, __u32);
+	__type(value, struct __perf_event_attr);
+} attr_heap SEC(".maps");
+
+#define VM_EXEC	      0x00000004
+#define MSG_OP_LOADER 26
+
+#define ATTR_BIT_MMAP	 (1ULL << 8)
+#define ATTR_BIT_MMAP2	 (1ULL << 23)
+#define ATTR_BIT_BUILDID (1ULL << 34)
+
+__attribute__((section(("kprobe/perf_event_mmap_output")), used)) int
+loader_kprobe(struct pt_regs *ctx)
+{
+	struct perf_mmap_event *mmap_event;
+	struct execve_map_value *curr;
+	struct task_struct *current;
+	struct vm_area_struct *vma;
+	struct msg_loader *msg;
+	unsigned long vm_flags;
+	struct perf_event *pe;
+	__u64 *id_map, id_pe;
+	const char *path;
+	size_t total;
+	int tgid;
+	long len;
+
+	msg = map_lookup_elem(&loader_heap, &(__u32){ 0 });
+	if (!msg)
+		return 0;
+
+	pe = (struct perf_event *)PT_REGS_PARM1_CORE(ctx);
+
+	/* Make sure it's our event that triggered perf_event_mmap_output,
+	 * to have all the needed info.
+	 */
```

Comment:
OK so to make sure I understand correctly:
 * we setup the perf monitoring from user-space
 * we get the perf event id from user-space and write it in the bpf map
 * we hook into the perf event output, and check whether the id of the event is the one we had set up
 * if it is, we get what we need from the event and write it to the event output
 
 And from what I remember from an offline discussion, we need to hook into the event because it's not easy to read the event generated in our main loop.

Reply:
> OK so to make sure I understand correctly:
> 
>     * we setup the perf monitoring from user-space
> 
>     * we get the perf event id from user-space and write it in the bpf map
> 
>     * we hook into the perf event output, and check whether the id of the event is the one we had set up
> 
>     * if it is, we get what we need from the event and write it to the event output
> 

exactly, I'll put this description into comments ;-) thanks

> And from what I remember from an offline discussion, we need to hook into the event because it's not easy to read the event generated in our main loop.

yes, we can't rely on perf events as is, because under kubernetes the perf subsystem
converts pid/tid values into perf event owner's namespace.. so we get zero pid/tid
for all processes executed outside tetragon namespace